### ISSUE TITLE
Add `result type` for Expr

### ISSUE THREAD

**Issue by jackwener** [apache/arrow-datafusion#6261]
### Is your feature request related to a problem or challenge?

In the current case, when we go to get the `result type` of Expr, we don't save it.
It also means that if we get the `result type` multiple times, it may require multiple runtime evaluate.

BTW, this feature is also related with `Type Coercion`, we should/must get `result type` after `Type Coercion`, because it may be invalid (need coerced type).

**Comment by jackwener** [comment]
After consideration, I think we just need to add `Result Type` for `BinaryExpr` `Case When` (complex Expr)... 
It's enough.

BUT, If there are big-refactorings in the future, we may also be able to add Result Type to all Expr.

**Comment by alamb** [comment]
I think it would be fine to add result type to complex expressions if we think it consumes significant amounts of time -- however, I think the biggest problem with coercion at the moment is that it is overly complicated (though your recent work to separate it out into argument coercion and result type has helped a lot in this area)

So my preference is to avoid adding more complexity (aka potentially inconsistencies with get_result_type) unless we see some evidence that this is actually consuming significant time

**Comment by liukun4515** [comment]
agree with @jackwener 

I think  type of expr should not be changed in the physical phase and all of type coercion should be done in the logical phase, but in current codebase, we can't do this.

The current implementation of the logical expr is `enum` and each type of logical lacks the result type , which is not friendly to achieve above thought.

In other SQL system, the logical expr is import part, each expr should has its input type and output type, and the output type need to be calculated from input types with operation. Once the common type and result are calculated in the logical phase, any changes for the type of expr should be forbidden.

As the comments pointed by @jackwener, some expr can't be evaluated many times. For example,

`old_expr = Add(left : decimal(10,2),right: decimal(10,2))`
after evaluating the first time:
the common type is decimal(11,2)
left casted from decimal(10,2) to decimal(11,2)
right casted from decimal(10,2) to decimal(11,2)
the result type is decimal(11,2)

and will get the `new_expr = Add(left: cast(decimal(10,2) to decimal(11,2)), right:cast(decimal(10,2) to decimal(11,2))`
If we evaluate the `new_expr` again, the new result will be generated.

@alamb

**Comment by alamb** [comment]
Makes sense @liukun4515  -- I agree it sounds like a good step forward to store the type of exprs.

> Once the common type and result are calculated in the logical phase, any changes for the type of expr should be forbidden.

Yes, this makes a lot of sense to me

@viirya  I think ran into some possibly related problem when working on decimal multiplication that required recasting in physical expressions. Would we be able to avoid such a case like:

https://github.com/apache/arrow-datafusion/blob/8a25953182b361a58c56384129bf57f32aa2dbb1/datafusion/physical-expr/src/expressions/binary.rs#L674-L688

**Comment by mingmwang** [comment]
> agree with @jackwener
> 
> I think type of expr should not be changed in the physical phase and all of type coercion should be done in the logical phase, but in current codebase, we can't do this.
> 
> The current implementation of the logical expr is `enum` and each type of logical lacks the result type , which is not friendly to achieve above thought.
> 
> In other SQL system, the logical expr is import part, each expr should has its input type and output type, and the output type need to be calculated from input types with operation. Once the common type and result are calculated in the logical phase, any changes for the type of expr should be forbidden.
> 
> As the comments pointed by @jackwener, some expr can't be evaluated many times. For example,
> 
> `old_expr = Add(left : decimal(10,2),right: decimal(10,2))` after evaluating the first time: the common type is decimal(11,2) left casted from decimal(10,2) to decimal(11,2) right casted from decimal(10,2) to decimal(11,2) the result type is decimal(11,2)
> 
> and will get the `new_expr = Add(left: cast(decimal(10,2) to decimal(11,2)), right:cast(decimal(10,2) to decimal(11,2))` If we evaluate the `new_expr` again, the new result will be generated.
> 
> @alamb



> agree with @jackwener
> 
> I think type of expr should not be changed in the physical phase and all of type coercion should be done in the logical phase, but in current codebase, we can't do this.
> 
> The current implementation of the logical expr is `enum` and each type of logical lacks the result type , which is not friendly to achieve above thought.
> 
> In other SQL system, the logical expr is import part, each expr should has its input type and output type, and the output type need to be calculated from input types with operation. Once the common type and result are calculated in the logical phase, any changes for the type of expr should be forbidden.
> 
> As the comments pointed by @jackwener, some expr can't be evaluated many times. For example,
> 
> `old_expr = Add(left : decimal(10,2),right: decimal(10,2))` after evaluating the first time: the common type is decimal(11,2) left casted from decimal(10,2) to decimal(11,2) right casted from decimal(10,2) to decimal(11,2) the result type is decimal(11,2)
> 
> and will get the `new_expr = Add(left: cast(decimal(10,2) to decimal(11,2)), right:cast(decimal(10,2) to decimal(11,2))` If we evaluate the `new_expr` again, the new result will be generated.
> 
> @alamb

@liukun4515 @jackwener 

In the current latest implementation,  this logic is moved out from the logical plan to physical plan.   That means evaluating multiple times will generate idempotent result.  I like the current implementation made by @viirya.
